finat/fiat_elements.py

```diff
@@ -173,6 +176,119 @@ def point_evaluation(self, order, refcoords, entity=None):
+        # order form a vector required for correct contraction with Q. Will
+        # become a FInAT.PointSet later.
+        # x_key = numpy.round(x_k, x_key_decimals) such that
+        # numpy.allclose(x_k, pt_hash, atol=1*10**-dec) == true
+        x = {}
+        x_key_decimals = 12
+        x_key_atol = 1e-12  # = 1*10**-dec
+
+        #
+        # BUILD Q TENSOR
+        #
+
+        # FIXME: The below loop is REALLY SLOW for BDM - Q and x should just be output as the dual basis
+
+        last_shape = None
+        self.Q_is_identity = True  # TODO do this better
+
+        fiat_dual_basis = self._element.dual_basis()
+        # i are rows of Q
+        for i, dual in enumerate(fiat_dual_basis):
+            point_dict = dual.get_point_dict()
+            for j, (x_j, tuples) in enumerate(point_dict.items()):
+                x_j = PointSet((x_j,))  # TODO doesn't need to be PointSet
+                # get weights q_j for point x_j from complicated FIAT dual data
+                # structure
+                weight_dict = {cmp: weight for weight, cmp in tuples}
+                if len(self._element.value_shape()) == 0:
+                    q_j = np.array(weight_dict[tuple()])
+                else:
+                    q_j = np.zeros(self._element.value_shape())
+                    for key, item in weight_dict.items():
+                        q_j[key] = item
+
+                # Ensure all weights have the same shape
+                if last_shape is not None:
+                    assert q_j.shape == last_shape
+                last_shape = q_j.shape
+
+                # Create key into x dict
+                x_key = np.round(x_j.points, x_key_decimals)
+                assert np.allclose(x_j.points, x_key, atol=x_key_atol)
+                x_key = tuple(x_key.flatten())
+
+                # Get value and index k or add to dict. k are the columns of Q.
+                try:
+                    x_j, k = x[x_key]
+                except KeyError:
+                    k = len(x)
+                    x[x_key] = x_j, k
+
+                # q_j may be tensor valued
+                it = np.nditer(q_j, flags=['multi_index'])
+                for q_j_entry in it:
+                    Q[(i, k) + it.multi_index] = q_j_entry
+                    if len(set((i, k) + it.multi_index)) > 1:
+                        # Identity has i == k == it.multi_index[0] == ...
+                        # Since i increases from 0 in increments of 1 we know
+                        # that if this check is not triggered we definitely
+                        # have an identity tensor.
+                        self.Q_is_identity = False
+
+        #
+        # CONVERT Q TO gem.Literal (TODO: should be a sparse tensor)
+        #
+
+        # temporary until sparse literals are implemented in GEM which will
+        # automatically convert a dictionary of keys internally.
+        Q = gem.Literal(sparse.as_coo(Q).todense())
```

Comment:
Now just need to find a neat way of getting shape...

Reply:
My suggestion
```python
            ndims = len(list(Q.keys())[0])
            shape = tuple(max(Q.keys(), key=itemgetter(i))[i] for i in range(ndims))
            Qdense = np.zeros(shape)
            for idx, value in Q.items():
                Qdense[idx] = value
            Q = Qdense
```

Reply:
I think you need to add 1 to all the extents in the shape (your maximum index is say `n` but the shape needs to be `n+1`), since we're zero-indexed, not 1-indexed.

one-liner:

```
shape = tuple(s + 1 for s in map(max, zip(*Q.keys()))
```

Reply:
🤦  quite right

One day I will know the language well enough to be able to come up with one liners like that